jctools-core/src/main/java/org/jctools/queues/MpscProgressiveChunkedQueue.java

```diff
@@ -0,0 +1,704 @@
+{
+    long p0, p1, p2, p3, p4, p5, p6, p7;
+    long p10, p11, p12, p13, p14, p15, p16;
+}
+
+/**
+ * An MPSC array queue which starts at <i>initialCapacity</i> and grows unbounded in linked chunks.<br>
+ * Differently from {@link MpscUnboundedArrayQueue} it is designed to provide a better scaling when more
+ * producers are concurrently offering.
+ *
+ * @param <E>
+ * @author https://github.com/franz1981
+ */
+public class MpscProgressiveChunkedQueue<E> extends MpscProgressiveChunkedQueuePad5<E>
+    implements MessagePassingQueue<E>, QueueProgressIndicators
+{
+    private final int chunkMask;
+    private final int chunkShift;
+
+    public MpscProgressiveChunkedQueue(int chunkSize)
+    {
+        chunkSize = Pow2.roundToPowerOfTwo(chunkSize);
+        final AtomicChunk<E> first = new AtomicChunk(0, null, chunkSize);
+        soProducerBuffer(first);
+        soProducerChunkIndex(0);
+        consumerBuffer = first;
+        chunkMask = chunkSize - 1;
+        chunkShift = Integer.numberOfTrailingZeros(chunkSize);
+        soFreeBuffer(new AtomicChunk(AtomicChunk.NIL_CHUNK_INDEX, null, chunkSize));
+    }
+
+    private AtomicChunk<E> producerBufferOf(AtomicChunk<E> producerBuffer, long expectedChunkIndex)
+    {
+        long jumpBackward;
+        while (true)
+        {
+            if (producerBuffer == null)
+            {
+                producerBuffer = lvProducerBuffer();
+            }
+            final long producerChunkIndex = producerBuffer.lvIndex();
+            if (producerChunkIndex == AtomicChunk.NIL_CHUNK_INDEX)
+            {
+                //force an attempt to fetch it another time
+                producerBuffer = null;
+                continue;
+            }
+            jumpBackward = producerChunkIndex - expectedChunkIndex;
+            if (jumpBackward >= 0)
+            {
+                break;
+            }
+            //try validate against the last producer chunk index
+            if (lvProducerChunkIndex() == producerChunkIndex)
+            {
+                producerBuffer = appendNextChunk(producerBuffer, producerChunkIndex, chunkMask + 1);
+            }
+            else
+            {
+                producerBuffer = null;
+            }
+        }
+        for (long i = 0; i < jumpBackward; i++)
+        {
+            //prev cannot be null, because is being released by index
+            producerBuffer = producerBuffer.lpPrev();
+            assert producerBuffer != null;
+        }
+        assert producerBuffer.lvIndex() == expectedChunkIndex;
+        return producerBuffer;
+    }
+
+    private AtomicChunk<E> appendNextChunk(AtomicChunk<E> producerBuffer, long chunkIndex, int chunkSize)
+    {
+        assert chunkIndex != AtomicChunk.NIL_CHUNK_INDEX;
+        final long nextChunkIndex = chunkIndex + 1;
+        //prevent other concurrent attempts on appendNextChunk
+        if (!casProducerChunkIndex(chunkIndex, nextChunkIndex))
```

Comment:
If we want a much more robust code here it can be replaced with 
`casProducerChunkIndex(chunkIndex, AtomicChunk.NIL_CHUNK_INDEX)` and a 
`soProducerChunkIndex(nextChunkIndex)` at the end IF no OOM error will happen :)
If a OOM error will happen it would re-set the original value `soProducerChunkIndex(nextChunkIndex)`,
leaving other threads able to re-try to allocate a new chunk or (if lucky enough) that the consumer
will provide a recycled chunk in time :)
I see that the other linked queues are not providing protection vs OOM, that's not strictly necessary, but 
given that are using `cas` it shouldn't complicate that much the impl